From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Thu, 1 Dec 2022 00:12:14 +0100
Subject: [PATCH] Update boss bar within tick

License: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)
Gale - https://galemc.org

This patch is based on the following mixin:
"net/caffeinemc/mods/lithium/mixin/ai/raid/RaidMixin.java"
By: Angeline <jellysquid3@users.noreply.github.com>
As part of: Lithium (https://github.com/CaffeineMC/lithium-fabric)
Licensed under: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)

diff --git a/net/minecraft/world/entity/raid/Raid.java b/net/minecraft/world/entity/raid/Raid.java
index 6e8c1a2863ac6e5137a26815ecf5142f0fcc9893..27234627933ccaa70b2c51a792efb2e68104a84f 100644
--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -109,6 +109,7 @@ public class Raid {
     private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry PDC_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
     public final org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(PDC_TYPE_REGISTRY);
     // Paper end
+    private boolean isBarDirty; // Gale - Lithium - update boss bar within tick
 
     public Raid(int id, ServerLevel level, BlockPos center) {
         this.id = id;
@@ -263,6 +264,12 @@ public class Raid {
     }
 
     public void tick() {
+        // Gale start - Lithium - update boss bar within tick
+        if (this.isBarDirty) {
+            this.updateBossbarInternal();
+            this.isBarDirty = false;
+        }
+        // Gale end - Lithium - update boss bar within tick
         if (!this.isStopped()) {
             if (this.status == Raid.RaidStatus.ONGOING) {
                 boolean flag = this.active;
@@ -588,6 +595,12 @@ public class Raid {
     }
 
     public void updateBossbar() {
+        // Gale start - Lithium - update boss bar within tick
+        this.isBarDirty = true;
+    }
+
+    private void updateBossbarInternal() {
+        // Gale end - Lithium - update boss bar within tick
         this.raidEvent.setProgress(Mth.clamp(this.getHealthOfLivingRaiders() / this.totalHealth, 0.0F, 1.0F));
     }
 
